x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/index.tsx

```diff
@@ -53,4 +53,6 @@ const CompareFieldsTableComponent: React.FC<Props> = ({
 
 CompareFieldsTableComponent.displayName = 'CompareFieldsTableComponent';
 
-export const CompareFieldsTable = React.memo(CompareFieldsTableComponent);
+export const CompareFieldsTable = React.memo(
+  CompareFieldsTableComponent
+) as typeof CompareFieldsTableComponent;
```

Comment:
For reviewers:

This is the cheapast fix for `React.memo` generic type forwarding. `React.memo` type layer is insignificant anyways so this doesn't affect type safety.

Reply:
From what I understand, the `T` type parameter has been introduced to allow different subtypes for `enrichedFieldMetadata` and `getTableColumns`. But is the benefit worth all these "type gymnastics"? What's the practical difference between this approach, and using the generic `EnrichedFieldMetadata` type directly for those props?

Reply:
Good catch. 

My intention was to make it more consistent with the rest of the changes and a little less confusing for getCustom... getEcs... getIncompatible... methods type-wise, because they are not supposed to return a wide type, they are supposed to return what their name explicitly mentions, it will be a problem the moment you try to access specific props that are only present in explicit type. But right now it's not happening and complicates the review. So I will undo these changes. 

Thanks.

Reply:
done

Reply:
oh and while undoing the changes @semd I came across a perfect example of why this change was actually pretty beneficial.

https://github.com/elastic/kibana/blob/8cd4858d2fbde97d3513feb58ee07cca304898cb/x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/compare_fields_table/get_incompatible_mappings_table_columns/index.tsx#L44

Here this assumes that `x` is of type `EnrichedFieldMetadata`. Which means every new logic you add here or even a mere check for `x.type` will create a type error here, you will have to needlessly narrow it down, even tho practically this function will always only accept incompatible fields that DO have `type` on them as opposed to custom fields. So by not narrowing down the type here (via generics in the accepting component) you are creating a situation when impossible states will happen, because TS will force them to check or assert, both of which is a bad and unnecessary. they will also have to accompany this with a needless test most of which I cleared in this PR. Let me know if this makes sense.

Reply:
https://github.com/elastic/kibana/assets/1625373/75aaca3d-1fdd-4f83-aa7a-9af20e1e93f5

Reply:
Okay, I see. It's funny because TS does not complain about `enrichedFieldMetadata` it detects those types are inclusive, it only complains about the `getTableColumns` I guess that's because it is a generic type. TS doing TS things....

If you want, there's another approach. Composing those components would allow us to forget about column definitions as a prop. Creating a separate component for `CompareInvalidFieldsTable`, `CompareECSFieldsTable`, `CompareCustomFieldsTable`, and `CompareSameFamilyFieldsTable`. They all would compose the current `CompareFieldsTable`, and each of them would have their own table columns definition inside the component, only receiving props for the _title_ and _fieldMetadata_. This way we also centralize all the table column definitions.

Afaik, React recommends "composing" over "extending", so this is the standard way I deal with this kind of things.
However, your first approach is also a good solution, this is not a "requested change" it's just a suggestion you can consider if you want.

Reply:
> Okay, I see. It's funny because TS does not complain about enrichedFieldMetadata it detects those types are inclusive, it only complains about the getTableColumns I guess that's because it is a generic type. TS doing TS things....

Yes because here 
![image](https://github.com/elastic/kibana/assets/1625373/0392070d-cbfc-4833-ad7d-e3e50080d1ca)

`EnrichedFieldMetadata` is including `EcsBasedFieldMetadata` and `CustomFieldMetadata`. So by passing narrower type into wider type you get no errors. And that's a logical and expected outcome. Because TS assumes that you just want to be able to pass in subtypes and that's exactly what we want here.

The problem happens inside of one of each of the methods like 
![image](https://github.com/elastic/kibana/assets/1625373/70a16f86-2915-4ca2-bcfb-9a3f853965d5)

Which is also very expected if you tell TS that "hey I will have `EnrichedFieldMetadata`" in here, so when accessing `x.<field>`, it breaks down as it SHOULD. This is a safety mechanism which is to be expected for you to not accidentally try to access a field that doesn't exist.

That's why you need to tell TS exactly what you expect to work with or as close as possible so it can help you better avoid needless checks and needless tests for those checks.

Long story short, TS is good here, it helps and defends against impossible states and confusion. The most important thing is to keep your root types healthy and as close to implementation as possible. If that's done TS will do its job of making sure all those impossible tests and checks never happen again in the future.

> If you want, there's another approach. 

Firstly I would suggest we start with why the particular composition approach (which I don't fully understand from your description unfortunately) will work better than existing suggestion from perspective type changes? Meaning what is the unfixable or cumbersome problem that current columned approach is introducing that would otherwise not  happen? Because if the problem is the introduction of component generic types, I'd argue it's not really a problem, but just another valid use of types with components. The `typeof Component` thing is just a limitation of react types nothing more, it doesn't make the entire approach invalid and needing to be completely refactored.

But I am open to your arguments and reasoning about the potential change, let's catch up in person during my implementation of the UI changes.

And now I will resolve this discussion and return my changes. Thanks for your input.

Reply:
Zoomed with Karen, I am convinced that the casting here was to support the generic type for CompareFieldsTable, so we can narrow dow EnrichedFieldMetadata to the specific types when rendered in each tab. Making the types specific can reduce redundant checks in the future and therefor reduce the test cases needed to cover those logic. I'm fine with the change here.

Reply:
thanks @angorayc also as mentioned in our discussion if this use case becomes more popular we can always extract it out into util. And btw just noticed it's already used in kibana in other teams as well  ![image](https://github.com/elastic/kibana/assets/1625373/3a2406f7-1fad-4560-b297-d32e2ccb7152)

Reply:
Yeah, I am also fine, I am not saying the current approach is wrong. I am just saying that this approach is something we don't usually do, this would be the first component I see doing that. That's why I would lean towards composing rather than making generic types, but I think both are okay. The behavior would be pretty much the same. But instead of doing this:
```
<CompareFieldsTable
    title="foo"
    enrichedFieldMetadata={incompatibleFieldsMetadata} // generic T = EcsFieldMetadata
    getTableColumns(getIncompatibleFieldsTableColumns} // Array<EuiTableFieldDataColumnType<EcsFieldMetadata>>
/>
```

We would do this:
```
<CompareIncompatibleFieldsTable
    title="foo"
    fieldsMetadata={incompatibleFieldsMetadata} // explicit EcsFieldMetadata type
/>
``` 

I like it because it's more explicit and simple, and there is some functionality (column definition) we could encapsulate inside each component. 
But as you want, I am okay with both.